RNAseq_AS Pipeline
================

The Alternative Splicing (AS) pipeline is based on the standard MUGQIC RNA-SEQ pipeline.
It uses RNA-SEQ to align reads and then uses Cufflinks, [Vast Tools](https://github.com/vastgroup/vast-tools), 
[MISO](http://genes.mit.edu/burgelab/miso/) and [JunctionSeq](https://github.com/hartleys/JunctionSeq) to perform differential analysis.

The RNAseq_AS pipeline requires a [readset file](https://bitbucket.org/mugqic/mugqic_pipelines/src#markdown-header-readset-file) which defines the readsets to be used in the pipeline. 
It also requires a [design file](https://bitbucket.org/mugqic/mugqic_pipelines/src#markdown-header-design-file) which will be used to define group comparison in the differential analyses.

Finally, a summary html report is automatically generated by the pipeline at the end of the analysis.
This report contains description of the sequencing experiment as well as a detailed presentation of 
the pipeline steps and results. Various Quality Control (QC) summary statistics and differential analysis 
results are included in the report. The report includes also the main references of the software tools 
and methods used during the analysis, together with the full list of parameters that have been passed 
to the pipeline main script.

Running RNAseq_AS Pipeline
-----

Refer to the MUGQIC, Vast Tools, MISO and JunctionSeq documentation pages for information on how to install each tool.

The RNAseq_AS pipeline has the same usage requirements as the RNAseq pipeline. Example setupenv.sh and generate-qsub.sh scripts can be found in `mugqic_pipelines/examples/rnaseq`. 

To run the pipeline, follow these steps:

1) Create a folder to hold your files.

2) Create a setup script in your folder (eg. setupenv.sh) and call `source setupenv.sh` in the command line.

setupenv.sh should look like this:
setupenv.sh
--- 
```
#!/bin/bash
# Set up the environment for GenAP
module load mugqic-pipelines/2.2.0
export PYTHONPATH=/path-to-your-folder/:$PYTHONPATH
```
3) Create the readset, design and config files as specified on the MUGQIC page.

4) Create and run a generate-qsub file. Here is an example generate-qsub.sh:
generate-qsub.sh
--- 
```
#!/bin/bash

/path-to-mugqic-code/pipelines/rnaseq_AS/RNAseq_AS.py \
 -s 1-23 \
 -o /path-to-your-folder/output \
 -j pbs \
 -l debug \
 -d rnaseq_AS.design \
 -r rnaseq_AS.readset \
 -c RNAseq_AS.ini 1> qsub.sh 2> debug.log

chmod +x qsub.sh

```
Check your debug.log and qsub.sh file for errors. Most errors are likely due to an incorrect setting in your configuration file.

5) Run your qsub.sh file. (ie. enter `./qsub.sh` in the command line)

Report Generation
-----

As the pipeline is running, it automatically creates the report in parts (based on the step). However, a further step must be taken to combine the segments into one report.

At any time during the pipeline processing, you can run the same pipeline command and add the option --report for the compilation of all the reports already created into a single HTML document (`output/report/index.html`).

To create the report, follow these steps:

1) Modify your generate-qsub.sh and add the `--report` parameter.
Example
---
```
#!/bin/bash

/path-to-mugqic-code/pipelines/rnaseq_AS/RNAseq_AS.py \
 -s 1-26 \
 -o /path-to-your-folder/output \
 -j pbs \
 -l debug \
 -d rnaseq_AS.design \
 -r rnaseq_AS.readset \
 -c RNAseq_AS.ini \
 --report 1> qsub.sh 2> debug.log

chmod +x qsub.sh
```
2) Run the generate-qsub.sh script once more, with the report parameter added.

3) Run your generated qsub.sh script.

Usage
-----
```
usage: RNAseq_AS.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
                 [-o OUTPUT_DIR] [-j {pbs,batch}] [-f] [--report] [--clean]
                 [-l {debug,info,warning,error,critical}] [-d DESIGN]
                 [-r READSETS] [-v]

Version: 2.2.0

For more documentation, visit the mugqic website: https://bitbucket.org/mugqic/mugqic_pipelines/

optional arguments:
  -h                    show this help message and exit
  --help                show detailed description of pipeline and steps
  -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                        config INI-style list of files; config parameters are
                        overwritten based on files order
  -s STEPS, --steps STEPS
                        step range e.g. '1-5', '3,6,7', '2,4-8'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        output directory (default: current)
  -j {pbs,batch}, --job-scheduler {pbs,batch}
                        job scheduler type (default: pbs)
  -f, --force           force creation of jobs even if up to date (default:
                        false)
  --report              create 'pandoc' command to merge all job markdown
                        report files in the given step range into HTML, if
                        they exist; if --report is set, --job-scheduler,
                        --force, --clean options and job up-to-date status are
                        ignored (default: false)
  --clean               create 'rm' commands for all job removable files in
                        the given step range, if they exist; if --clean is
                        set, --job-scheduler, --force options and job up-to-
                        date status are ignored (default: false)
  -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                        log level (default: info)
  -d DESIGN, --design DESIGN
                        design file
  -r READSETS, --readsets READSETS
                        readset file
  -v, --version         show the version information and exit

Steps:
------
1- picard_sam_to_fastq
2- trimmomatic
3- merge_trimmomatic_stats
4- star
5- picard_merge_sam_files
6- bam_hard_clip
7- cufflinks
8- cuffmerge
9- cuffquant
10- cuffdiff
11- cuffnorm
12- miso_index
13- miso_paired_end
14- miso_psi
15- miso_summarize
16- miso_diff
17- miso_plot
18- bam_to_fastq
19- vast_tools_align
20- vast_tools_combine
21- vast_tools_diff
22- vast_tools_plot
23- jsctseq_raw_counts
24- jctseq_make_gff
25- jctseq_diff_prep
26- jctseq_diff

```
1- picard_sam_to_fastq
----------------------
Convert SAM/BAM files from the input readset file into FASTQ format
if FASTQ files are not already specified in the readset file. Do nothing otherwise.

2- trimmomatic
--------------
Raw reads quality trimming and removing of Illumina adapters is performed using [Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic).
If an adapter FASTA file is specified in the config file (section 'trimmomatic', param 'adapter_fasta'),
it is used first. Else, 'Adapter1' and 'Adapter2' columns from the readset file are used to create
an adapter FASTA file, given then to Trimmomatic. For PAIRED_END readsets, readset adapters are
reversed-complemented and swapped, to match Trimmomatic Palindrome strategy. For SINGLE_END readsets,
only Adapter1 is used and left unchanged.

This step takes as input files:

1. FASTQ files from the readset file if available
2. Else, FASTQ output files from previous picard_sam_to_fastq conversion of BAM files

Please note that this step is optional as the STAR step will do its own trimming.

3- merge_trimmomatic_stats
--------------------------
The trim statistics per readset are merged at this step.

Please note that this step is optional as the STAR step will do its own trimming.

4- star
-------
The filtered reads are aligned to a reference genome. The alignment is done per readset of sequencing
using the [STAR](https://code.google.com/p/rna-star/) software. It generates a Binary Alignment Map file (.bam).

This step takes as input files:

1. Trimmed FASTQ files if available
2. Else, FASTQ files from the readset file if available
3. Else, FASTQ output files from previous picard_sam_to_fastq conversion of BAM files

5- picard_merge_sam_files
-------------------------
BAM readset files are merged into one file per sample. Merge is done using [Picard](http://broadinstitute.github.io/picard/).

6- bam_hard_clip
-----------------
Generate a hardclipped version of the bam for the toxedo suite which doesn't support this official sam feature.

This step is only needed if you are running cufflinks.

7- cufflinks
-------------
Compute RNA-Seq data expression using [cufflinks](http://cole-trapnell-lab.github.io/cufflinks/cufflinks/).
Warning: It needs to use a hard clipped bam file while Tuxedo tools do not support official soft clip SAM format

This step is only needed if you are running cufflinks.

8- cuffmerge
-------------
Merge assemblies into a master transcriptome reference using [cuffmerge](http://cole-trapnell-lab.github.io/cufflinks/cuffmerge/).

This step is only needed if you are running cufflinks.

9- cuffquant
-------------
Compute expression profiles (abundances.cxb) using [cuffquant](http://cole-trapnell-lab.github.io/cufflinks/cuffquant/).
Warning: It needs to use a hard clipped bam file while Tuxedo tools do not support official soft clip SAM format

This step is only needed if you are running cufflinks.

10- cuffdiff
------------
[Cuffdiff](http://cole-trapnell-lab.github.io/cufflinks/cuffdiff/) is used to calculate differential transcript expression levels and test them for significant differences.

This step is only needed if you are running cufflinks.

11- cuffnorm
------------
Global normalization of RNA-Seq expression levels using [Cuffnorm](http://cole-trapnell-lab.github.io/cufflinks/cuffnorm/).

This step is only needed if you are running cufflinks.

12- miso_index
------------
Creates an index of gff annotations for MISO.

Important Config Settings:

1. `input_gff_directory` is the full path to the directory containing the gffs to be indexed for miso_psi.
2. `gff_assembly` is the assembly in the gff name (ie. `hg19` or `mm9`)
3. `input_gff_prefixes` is a space separated list with the prefixes to the gff files. (ie. `A3SS A5SS MXE RI SE`)
This step will combine `{input_gff_directory}/{input_gff_prefix}.{gff_assembly}.gff3` and index each prefix. Note: It will automatically determine whether to use `.gff3` or `.nochr.gff3`.

This step is only needed if you are running MISO.

13- miso_paired_end
------------
Computes insert length distribution and its statistics for the next step by running `pe_utils --compute-insert-len`. This step is only needed for paired end reads, 
and will run if `library_type=paired` in the `miso_psi` section of the config file. This step will be skipped if `library_type=single`.

Important Config Settings:

1. `input_gff_file` is the full path to the gff to be used for this step.
2. `min_exon_size` sets the minimum required length for exons.
3. `other_options` can be any other options for `pe_utils --compute-insert-len`.

This step is only needed if you are running MISO.

14- miso_psi
------------
Estimates the expression level of a set of annotated isoforms or events by calculating PSI values.
This function will automatically calculate read length and whether to run in paired end mode.

Important Config Settings:

1. `library_type` should be `paired` or `single`.
2. `other_options` can be any other options for `miso --run`.

This step is only needed if you are running MISO.

15- miso_summarize
------------
Summarizes the output from the previous step and provides confidence intervals for the PSI values. 

Important Config Settings:

1. `other_options` can be any other options for `summarize_miso --summarize-samples`.

This step is only needed if you are running MISO.

16- miso_diff
------------
Computes Bayes factor to determine how likely the exon or isoform is differentially expressed between samples. 
Note: `compare_miso --compare-samples` can only compare 2 samples at a time. If more than one control or treatment is set for a contrast, MISO will compare the first control and treament samples.

Important Config Settings:

1. `other_options` can be any other options for `compare_miso --compare-samples`.

This step is only needed if you are running MISO.

17- miso_plot
------------
Uses [Sashimi Plot](http://miso.readthedocs.io/en/fastmiso/sashimi.html) to plot the results from MISO.

Important Config Settings:

1. `plot_type` must be `--plot-insert-len`, `--plot-bf-dist` or `--plot-event` or any combination of the 3. NOTE: `--plot-insert-len` and `--plot-bf-dist` may not working due to Sashimi bugs.
2. `event_names` should be the MISO ID (ie `chr4:53951933:53952063:+@chr4:53952354:53952422:+@chr4:53960654:53960763:+`) of each event to be plotted separated by a space. This option is only required for plot type `--plot-event`.
3. The options under the `miso_plot_settings` section are to customize the plots. Check the Sashimi Plot documentation for information on these options.

This step is only needed if you are running MISO.

18- bam_to_fastq
------------
Uses Tophat to change the aligned bam files into fastq files for Vast Tools to use.

This step is only needed if you are running Vast Tools.

19- vast_tools_align
------------
Outputs the alignment data in a suitable format for Vast Tools usage and computes PSI values.

Important Config Settings:

1. `species` must be `Hsa` (human), `Mmu` (mouse) or `Gga` (chicken)
2. `other_options` can be any other options for `vast-tools align`.

This step is only needed if you are running Vast Tools.

20- vast_tools_combine
------------
Combines the data from each sample into one table.

Important Config Settings:

1. `other_options` can be any other options for `vast-tools combine`.

This step is only needed if you are running Vast Tools.

21- vast_tools_diff
------------
Uses the `vast-tools diff` function to determine how likely an exon is differentially expressed between samples.

Important Config Settings:

1. `other_options` can be any other options for `vast-tools diff`.

This step is only needed if you are running Vast Tools.

22- vast_tools_plot
------------
Plots the results from Vast Tools.

Important Config Settings:

1. `significant_events` should be the Vast Tools ID (ie `HsaEX0000000`) of each event to be plotted separated by a space.
2. `other_options` can be any other options for `vast-tools plot`.

This step is only needed if you are running Vast Tools.

23- jctseq_raw_counts
------------
Uses [QoRTs](http://hartleys.github.io/QoRTs/index.html) to calculate gene-level and splice-junction-level counts. 

Important Config Settings:

1. `library_type` should be `single` or `paired`. QoRTs will run in `--singleEnded` mode or not depending on this value for this step and the next.
2. `strand` should be `stranded` or `unstranded. QoRTs will run in `--stranded` mode or not for this step and the next.
2. `QoRTs_other_options` can be any other options for `java -jar QoRTs.jar QC`.

This step is only needed if you are running JunctionSeq.

24- jctseq_make_gff
------------
Uses QoRTs to generate an annotation file used by JunctionSeq. It uses the gtf specified in the `DEFAULT` section to create the gff file.

This step is only needed if you are running JunctionSeq.

25- jctseq_diff_prep
------------
Prepares the data for the next step.

Important Config Settings:

1. `nCores` is the number of cores that the will be used when `runJunctionSeqAnalyses` is called in the next step.

This step is only needed if you are running JunctionSeq.

26- jctseq_diff
------------
Tests for differential usage of exons and splice junctions. It will also plot the results of JunctionSeq.

This step is only needed if you are running JunctionSeq.
